[Draft] Add NanoBEIREvaluator

NohTow opened the pull request:
This PR adds [NanoBEIREvaluator](https://x.com/tomaarsen/status/1857434642569138243) to PyLate, an evaluator that compute various metrics over subsets of BEIR datasets, allowing to get a glance of the training behavior without taking too long to run.

The usage is really simple, it just requires to add `evaluator=evaluation.NanoBEIREvaluator()` to the Trainer.
The code is _very_ similar to the one of ST, it could basically directly leverage it with the proper similarity function, but since our encoding is asymmetrical and requires to pad the documents, I extended InformationRetrievalEvaluator to override the only function that need change and also made a NanoBeirEvaluator that use this.

To make things cleaner, I also fixed the similarity_fn_name (and the different process associated with it) to make MaxSim the default, which makes it nicer and also fix the issue of having "cosine" as the similarity function in the config.

Besides W&B tracking, this allows to add the evaluation to the model card, which is very handy for people releasing the models.

Edit: that is still a draft because I need to check for side effects, but the code should be ok already

raphaelsty commented:
Can you update the file `setup.py` and add the model_card_template.md to the package data ? It will allow users to download this and store it in the pylate folder when installing the package. 

```
setuptools.setup(
    ...
    package_data={"pylate": ["hf_hub/model_card_template.md"]},
    ...
)
```

raphaelsty commented on pylate/evaluation/nano_beir_evaluator.py:
```diff
@@ -0,0 +1,206 @@
+]
+
+
+dataset_name_to_id = {
```
You can rename this to 

`MAPPING_DATASET_NAME_TO_ID`

with uppercase

raphaelsty commented on pylate/evaluation/nano_beir_evaluator.py:
```diff
@@ -0,0 +1,206 @@
+    "touche2020": "zeta-alpha-ai/NanoTouche2020",
+}
+
+dataset_name_to_human_readable = {
```
You can rename this to

MAPPING_DATASET_NAME_TO_READABLE_FORMAT

with uppercase

raphaelsty commented on pylate/evaluation/nano_beir_evaluator.py:
```diff
@@ -0,0 +1,206 @@
+    The Evaluator will return the same metrics as the InformationRetrievalEvaluator (i.e., MRR, nDCG, Recall@k), for each dataset and on average.
+
+
+    Example:
```
This exemple docstring format won't be parsed by the documentation, you can use the standard format. Maybe we could use the example as unit test and rather import pylate

raphaelsty commented on pylate/evaluation/pylate_information_retrieval_evaluator.py:
```diff
@@ -0,0 +1,230 @@
+    Given a set of queries and a large corpus set. It will retrieve for each query the top-k most similar document. It measures
+    Mean Reciprocal Rank (MRR), Recall@k, and Normalized Discounted Cumulative Gain (NDCG)
+
+    Example:
```
The docstring format won't be parsed by our doc and it could be cool to create some unit tests in this docstring

raphaelsty commented on pylate/models/colbert.py:
```diff
@@ -752,6 +754,35 @@ def pool_embeddings_hierarchical(
+            Optional[str]: The name of the similarity function. Can be None if not set, in which case it will
+                default to "cosine" when first called.
+
+        Example:
```
The docstring format won't be parsed by our doc and it could be cool to create some unit tests in this docstring

raphaelsty commented on pylate/scores/similarity_functions.py:
```diff
@@ -0,0 +1,81 @@
+        """
+        Converts a similarity function name or enum value to the corresponding similarity function.
+
+        Args:
```
The docstring format won't be parsed by our doc

raphaelsty commented on pylate/scores/similarity_functions.py:
```diff
@@ -0,0 +1,81 @@
+        Returns:
+            list: A list of possible values for the SimilarityFunction enum.
+
+        Example:
```
The docstring format won't be parsed by our doc

NohTow commented:
I added some tests in the docstring and reworked them, removed some of the docs that are just ST extensions anyways.
Can you have a second look @raphaelsty?

raphaelsty commented:
LGTM !! NanoBeirEvaluator is an amazing feature
